### PR TITLE
feat: implement direct file upload

### PR DESCRIPTION
**Pull request by luka-nextcloud**
* Resolve: #833
* Description: Ability to import table with direct file upload.

[demo (2).webm](https://github.com/nextcloud/tables/assets/89908051/f90caa4b-ac01-43b7-8ced-0b093fdcb8ad)

### CODE REVIEW

**Review by blizzz — APPROVED** [review]
Looks good. Could not test fully, because of an unrelated error.

One nitpick: when I upload a beautiful (i.e. formatted cells) ods file for import with four rows, it also tries to import the empty rows, resulting in 16384 row creation errors. Might an issue of the IOFactory lib though, and unrelated to this PR.

Screenshot_20240215_190613

**Review by enjeck — COMMENTED** [review]
This is a cool feature :sunglasses: 

It's not immediately obvious to me how `Select a file` and `Upload a file` are different. I think we should rename the buttons to be consistent with what we have in other Nextcloud apps. For example, in collectives, we use `Upload from computer` and `Insert from Files`.
image

**Comment by juliushaertl** [comment]
@nextcloud/designers Maybe you can check this one and suggest how we should layout the different upload buttons?

**Comment by nimishavijay** [comment]
This is a super cool feature! :) General flow looks good! Some comments:

- [ ] Agreed with @enjeck that the buttons can be named "Upload from device" and "Insert from Files" :) 
- [ ] The placement of the buttons can also be improved. The text fields next to the buttons can be removed, the buttons can be on the same row and selected file name can be above the buttons
- [ ] In the "Create missing columns" section, the explanation text is cut off. It should wrap around, or can even be removed.
- [ ] The "Information" section currently has a lot of text, some of it can be removed (for example, the "The possible importing..." point seems very technical, would it cause any confusion if it was not there?) and some of it can be inserted in relevant places

So the end result will look something like this

image
image


- [ ] The text on the loading screen can also be reduced if it is not critical 
- [ ] Instead we can show the name of the file (ellipsised if it is too long)

image


- [ ] For the result screen, there can be an icon on the left of the heading indicating if there are any errors. If there are no errors, then it can be a checkmark and if there are errors it can be a warning 
- [ ] If there no errors, the last 2 rows can be hidden
- [ ] If there are errors, it could be indicated by an icon or emoji

```
✅ Imported from n26_transactions.csv

Found columns        4
Matching columns     0
Created columns      4
Inserted rows        0
```

```
⚠️ Imported from n26_transactions.csv

Found columns          4
Matching columns       0
Created columns        4
Inserted rows          0
Value parsing errors   1 ⚠️ 
Row creation errors    1 ⚠️      
```
What do you think? :)

File: src/modules/modals/Import.vue [hunk 1]
```diff
@@ -180,6 +198,15 @@ export default {
 			loading: false,
 			result: null,
 			waitForReload: false,
+			mimeTypes: [
+				'text/csv',
+				'application/vnd.ms-excel',
+				'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+				'application/xml',
+				'text/html',
```
[the remark this answers is not in the document]
Comment (luka-nextcloud) [reply src/modules/modals/Import.vue:211]: We are using phpspreasheet to parse import file. It does support xml and html file formats. (https://phpspreadsheet.readthedocs.io/en/latest/)

**Comment by luka-nextcloud** [comment]
[demo (3).webm](https://github.com/nextcloud/tables/assets/89908051/e3c50fdf-c0c5-41c3-aa32-1afd99aff8c6)

**Review by juliushaertl — APPROVED** [review]
Very nice 👍

**Comment by juliushaertl** [comment]
@blizzz Regarding your row errors mentioned in https://github.com/nextcloud/tables/pull/845#pullrequestreview-1883456697 any chance you have a copy of the file you tried? I could not reproduce that in any way

**Pull request merged**